nntrainer/layers/conv2d_layer.cpp

```diff
@@ -446,103 +461,89 @@ void Conv2DLayer::setProperty(const PropertyType type,
+    }
 
-  Tensor im2col_array;
-  if (channel_mode) {
-    im2col_array = Tensor(kdim.getFeatureLen(), out_height * out_width);
-  } else {
-    im2col_array =
-      Tensor(k_height * k_width, in.channel() * out_height * out_width);
-  }
-
-  if (pad_value == 0) {
-    im2col_array.setZero();
-  } else {
-    /// not reaching here, just preparing for non-zero pad_value
-    im2col_array.setValue(pad_value);
+    if (pad_value == 0) {
+      out.setZero();
+    } else {
+      /// not reaching here, just preparing for non-zero pad_value
+      out.setValue(pad_value);
+    }
   }
 
-  auto in_range = [](unsigned int virtual_pos, unsigned int pad,
-                     unsigned int actual_len) -> bool {
-    return pad <= virtual_pos && virtual_pos < (pad + actual_len);
-  };
+  float *out_data = out.getData();
 
   if (channel_mode) {
-    unsigned int h_stride_end = height - k_height;
-    unsigned int w_stride_end = width - k_width;
+    int h_stride_end = height - k_height - ph;
+    int w_stride_end = width - k_width - pw;
+
     /// get a patch, size of kernel
     /// hs is height_strided, ws is width_strided
-    unsigned int im_w = 0;
-    for (unsigned int hs = 0; hs <= h_stride_end; hs += mstride[0]) {
-      for (unsigned int ws = 0; ws <= w_stride_end; ws += mstride[1]) {
-        unsigned int im_h = 0;
-        unsigned int patch_height_end = k_height + hs;
-        unsigned int patch_width_end = k_width + ws;
-
-        for (unsigned int c = 0; c < channel; ++c) {
-          /// map the patch to a single line loopting through channel
-          for (unsigned int h = hs; h < patch_height_end; ++h) {
-            if (!in_range(h, ph, in.height())) {
-              im_h += k_width;
-              continue;
-            }
+    unsigned int owidth = out.width();
+    unsigned int base_im_w = 0;
+    for (int hs = -ph; hs <= h_stride_end; hs += mstride[0]) {
+      unsigned int base_im_h = 0;
+      int patch_height_end = k_height + hs;
+      /// map the patch to a single line looping through channel
+      for (unsigned int c = 0; c < channel; ++c) {
+        for (int h = hs; h < patch_height_end; ++h) {
+          if (h < 0 || in_height <= h) {
+            base_im_h += k_width;
+            continue;
+          }
 
-            for (unsigned int w = ws; w < patch_width_end; ++w) {
-              if (!in_range(w, pw, in.width())) {
+          unsigned int im_w = base_im_w;
+          for (int ws = -pw; ws <= w_stride_end; ws += mstride[1]) {
+            unsigned int im_h = base_im_h;
+            int patch_width_end = k_width + ws;
+
+            for (int w = ws; w < patch_width_end; ++w) {
+              if (w < 0 || in_width <= w) {
                 im_h++;
                 continue;
               }
-              float val = in.getValue(0, c, h - ph, w - pw);
-              im2col_array.setValue(0, 0, im_h, im_w, val);
+              out_data[im_w * owidth + im_h] = in.getValue(0, c, h, w);
               im_h++;
             }
```

Comment:
how about replace this loop with a simple memcpy? as the copy seems to be happening from contiguous one region to another.

`memcpy(out_data + im_w * owidth + im_h, in + in.getIndex(0,c,h,ws), appropriate size based on all the conditions).`
Current loop wont be optimized by compiler because of the if condition inside.

Reply:
Good idea, I'll try

Reply:
WIP, I'll leave this as an issue as optimized version is needed now